posts/coding-intentionally-in-bash-grains.md

```diff
@@ -0,0 +1,127 @@
+
+### First Option: Binary
+
+Binary is closer to the core of the exercise.  We could replace the F's with 64 1's.
+
+```bash
+bc <<< "ibase=2;1111111111111111111111111111111111111111111111111111111111111111"
+```
+
+More intentional, but, with it being so long and with us not speaking robot, a long, essentially uncountable string of 1's is maybe not an improvement.
+
+### Second Option: Brute-Force Calculation
+
+OK, so maybe we abandon the non-decimal numbering systems altogether.  Why don't we make the code match how we would total up the number of grains on a chessboard by hand, by counting the grains on each square?
+
+```bash
+total=0
+current_grains=1
+
+for square in {1..64}; do
+  total=$( bc <<< "$total + $current_grains" )
+  current_grains=$( bc <<< "$current_grains * 2" )
+done
+
+echo "$total"
+```
+
+That's a lot more readable and understandable.  Controls are shown for the number of squares on the chess board (although the 64 could be pulled out into a defined constant if need be), and the doubling effect is clearly shown.  I actually think this is better than the initial solution.
+
+However.
+
+It's slow.  Looping, adding, and repeatedly calling out to an external command?  They all stack up to a slow-ish run-time.  Now, is that a huge deal?  No.  If you're scripting this in Bash, you've probably already decided you don't have speed constraints.  But could it be better?  Yes.
+
+### Third Option: Direct Calculation
+
+In the math biz (that's what cool math folks call it), we've got a name for the process "take something and then multiply it by the same number a bunch of times": exponents.
+
+In fact, the number of grains on each square can be expressed as:
+
+```bash
+bc <<< "2^($square - 1)"
+# We subtract one from $square because square 1 has 1 (2^0) grains on it.
+```
+
+So, how do we add all these up without iterating?  There are several ways to think about it, but here are the two that seem to resonate with most people.
+
+Let's look at the first few squares again, but, this time, we'll *only* look at the running total *up to just before the square we're on*.
+
+```txt
+Square #      Power of 2      Value     Total Until Now
+       1             2^0          1                   0
+       2             2^1          2                   1
+       3             2^2          4                   3
+       4             2^3          8                   7
+       5             2^4         16                  15
+       6             2^5         32                  31
+       7             2^6         64                  63
+       8             2^7        128                 127
+```
+
+Are you seeing the pattern?  **The number of grains on a square is equal to the *sum of all the squares before it* plus one**.  *And*, we have a formula for the number of grains on a square.
```

Comment:
Getting people to discover this pattern themselves was one of my favorite things about the Grains exercise when I mentored on it a lot. I wonder if we can lead people to that delightful moment of discovery while reading the blog post.